packages/compartment-mapper/src/map-parser.js

```diff
@@ -0,0 +1,302 @@
+   */
+  function* getParserGenerator(
+    bytes,
+    specifier,
+    location,
+    packageLocation,
+    options,
+  ) {
+    /** @type {string} */
+    let language;
+    const extension = parseExtension(location);
+
+    if (
+      !extensionImpliesLanguage(extension) &&
+      has(languageForModuleSpecifier, specifier)
+    ) {
+      language = languageForModuleSpecifier[specifier];
+    } else {
+      language = languageForExtension[extension] || extension;
+    }
+
+    /** @type {string|undefined} */
+    let sourceMap;
+
+    if (has(transforms, language)) {
+      try {
+        ({
+          bytes,
+          parser: language,
+          sourceMap,
+        } = yield transforms[language](
+          bytes,
+          specifier,
+          location,
+          packageLocation,
+          {
+            // At time of writing, sourceMap is always undefined, but keeping
+            // it here is more resilient if the surrounding if block becomes a
+            // loop for multi-step transforms.
+            sourceMap,
+          },
+        ));
+      } catch (err) {
+        throw Error(
+          `Error transforming ${q(language)} source in ${q(location)}: ${err.message}`,
+          { cause: err },
+        );
+      }
+    }
+    if (!has(parserForLanguage, language)) {
+      throw Error(
+        `Cannot parse module ${specifier} at ${location}, no parser configured for the language ${language}`,
+      );
+    }
+    const { parse } = parserForLanguage[language];
+    return parse(bytes, specifier, location, packageLocation, {
+      sourceMap,
+      ...options,
+    });
+  }
+
+  /**
+   * @type {ParseFn}
+   */
+  const syncParser = (bytes, specifier, location, packageLocation, options) => {
+    const result = syncTrampoline(
+      getParserGenerator,
+      bytes,
+      specifier,
+      location,
+      packageLocation,
+      options,
+    );
+    if ('then' in result && typeof result.then === 'function') {
+      throw new TypeError(
+        'Sync parser cannot return a Thenable; ensure parser is actually synchronous',
+      );
+    }
+    return result;
+  };
+  syncParser.isSyncParser = true;
+
+  /**
+   * @param {Uint8Array} bytes
+   * @param {string} specifier
+   * @param {string} location
+   * @param {string} packageLocation
+   * @param {*} options
+   * @returns {Promise<ParseResult>}
+   */
+  const asyncParser = async (
+    bytes,
+    specifier,
+    location,
+    packageLocation,
+    options,
+  ) => {
+    return asyncTrampoline(
+      getParserGenerator,
+      bytes,
+      specifier,
+      location,
+      packageLocation,
+      options,
+    );
+  };
+
+  // if we have nothing in the moduleTransforms object, then we can use the syncParser.
+  if (keys(moduleTransforms).length === 0) {
+    transforms = syncModuleTransforms;
+    return syncParser;
+  }
```

Comment:
This is incorrect.

Just because `moduleTransforms` is empty doesn't mean that the parsers are sync.  We don't have enough information in `mapParsers()` to make that determination.

We do, however, have the information in `link()` (which calls `mapParsers()`).  I'm not sure what's the best way to refactor this though. Send `mapParsers()` a `sync` bit?  idk 

@naugtur

Reply:
fixed it here: https://github.com/endojs/endo/pull/2464